### PR TITLE
feat: [CDC-16636]: use analytic mongo node for graphql queries behind ff

### PR DESCRIPTION
**Pull request by inderschera**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30544)

### CODE REVIEW

File: 960-persistence/src/main/java/io/harness/mongo/MongoModule.java [hunk 1]
```diff
@@ -191,6 +194,23 @@ public AdvancedDatastore primaryDatastore(MongoConfig mongoConfig, @Named("morph
     return primaryDatastore;
   }
 
+  @Provides
+  @Named("analyticsDatabase")
+  @Singleton
+  public AdvancedDatastore getAnalyticsDatabase(MongoConfig mongoConfig, Morphia morphia) {
+    TagSet tags = new TagSet(new Tag("nodeType", "ANALYTICS"));
```
Comment (abhinav-harness) [inline 960-persistence/src/main/java/io/harness/mongo/MongoModule.java:201]: can we get this from config?

File: 960-persistence/src/main/java/io/harness/mongo/MongoModule.java [hunk 2]
```diff
@@ -191,6 +194,23 @@ public AdvancedDatastore primaryDatastore(MongoConfig mongoConfig, @Named("morph
...
+    final String mongoClientUrl = mongoConfig.getUri();
+    ReadPreference readPreference = ReadPreference.secondary(tags);
+    MongoClientURI uri = new MongoClientURI(mongoClientUrl,
+        MongoClientOptions.builder(MongoModule.getDefaultMongoClientOptions(mongoConfig))
+            .readPreference(readPreference));
+
+    MongoClient mongoClient = new MongoClient(uri);
+    AdvancedDatastore analyticalDataStore = (AdvancedDatastore) morphia.createDatastore(mongoClient, uri.getDatabase());
```
Comment (abhinav-harness) [inline 960-persistence/src/main/java/io/harness/mongo/MongoModule.java:209]: this wont be able to do init in on prem -- do you know what happens if we don't have the node with tag -- will be easy to check in local

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30544)* status: 9 file(s) to review. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[960-persistence/src/main/java/io/harness/mongo/MongoModule.java, line 201 at r1](https://reviewable.io/reviews/harness/harness-core/30544#-Mv2NBsYfWLL4vrV2fir-r1-201:-MxcZkmN3l9CZW8pLfIS:b-2maqyn) ([raw file](https://github.com/harness/harness-core/blob/d5c16dac0af5d5767c4d2ce0d8abe4bae9795707/960-persistence/src/main/java/io/harness/mongo/MongoModule.java#L201)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

can we get this from config?
</blockquote></details>

This is hard coded tags required for analytic mongo node. For other scenarios, tags are empty. 
We can probably take boolean from config.yml. If it's false we can get results from secondary node else we get those from analytic node.
___
*[960-persistence/src/main/java/io/harness/mongo/MongoModule.java, line 209 at r1](https://reviewable.io/reviews/harness/harness-core/30544#-Mv2NBsYfWLL4vrV2fir-r1-209:-Mxcb0Z44yylGAx5nynF:b-q5sggw) ([raw file](https://github.com/harness/harness-core/blob/d5c16dac0af5d5767c4d2ce0d8abe4bae9795707/960-persistence/src/main/java/io/harness/mongo/MongoModule.java#L209)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

this wont be able to do init in on prem -- do you know what happens if we don't have the node with tag -- will be easy to check in local 
</blockquote></details>

The query gets stuck in this case. We should not enable the feature flag for on-prem customers.
As said earlier, I will add boolean to config.yml. In case, the feature flag is enabled on on-prem, we will get results from secondary mongo.

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30544)* status: 15 file(s) to review. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[960-persistence/src/main/java/io/harness/mongo/MongoModule.java, line 201 at r1](https://reviewable.io/reviews/harness/harness-core/30544#-Mv2NBsYfWLL4vrV2fir-r1-201:-MxhEkON8GHb6HPM6DB3:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d5c16dac0af5d5767c4d2ce0d8abe4bae9795707/960-persistence/src/main/java/io/harness/mongo/MongoModule.java#L201)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

This is hard coded tags required for analytic mongo node. For other scenarios, tags are empty. 
We can probably take boolean from config.yml. If it's false we can get results from secondary node else we get those from analytic node.
</blockquote></details>

Done.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 7 of 9 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30544)* status: :shipit: complete! ✓

**Comment by inderschera** [comment]
trigger messagecheck

**Review by poojaSinghal — APPROVED** [review]
> Reviewed 6 of 9 files at r1, 1 of 1 files at r2, 6 of 7 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30544)* status: :shipit: complete! ✓

**Pull request merged**